### PR TITLE
Augment PluginDescription with fields used in remapping commands

### PR DESCRIPTION
**Pull request by vuil**
### What this PR does / why we need it

Introduce two additional fields in the PluginDescriptor, InvokedAs and SupportedContexttype, marked experimental are added to support plugin level remapping

### Which issue(s) this PR fixes

Fixes #

### Release note

```release-note
Introduce two experimental fields in the PluginDescriptor, InvokedAs and SupportedContexttype to support plugin-level remapping
```

### CODE REVIEW

**Review by anujc25 — APPROVED** [review]
LGTM.

File: plugin/types.go [hunk 1]
```diff
@@ -104,4 +104,19 @@ type PluginDescriptor struct {
 
 	// DefaultFeatureFlags is default featureflags to be configured if missing when invoking plugin
 	DefaultFeatureFlags map[string]bool `json:"defaultFeatureFlags,omitempty" yaml:"defaultFeatureFlags,omitempty"`
+
+	// InvokedAs provides a specific mapping to how any command provided by this plugin should be invoked as.
+	// If unset (which is equivalent to setting it to ["<PluginDescript.Name>"]), commands will typically be invocable
+	// with the Tanzu CLI using "<PluginDescriptor.Name> <command name> commandargs...."
+	// Can be used to specify additional levels in the command hiearchy via values with space-delimited parts.
+	// e.g. ["ops cluster"] implies plugin's command foo will be invoked by the Tanzu CLI using
```
Comment (anujc25) [inline plugin/types.go:112]: nit: As `ops` is an alias of the `operations` command, it would be better to use `operations` in the example here instead of `ops` just to avoid any confusion.

File: plugin/types.go [hunk 2]
```diff
@@ -104,4 +104,19 @@ type PluginDescriptor struct {
 
 	// DefaultFeatureFlags is default featureflags to be configured if missing when invoking plugin
 	DefaultFeatureFlags map[string]bool `json:"defaultFeatureFlags,omitempty" yaml:"defaultFeatureFlags,omitempty"`
+
+	// InvokedAs provides a specific mapping to how any command provided by this plugin should be invoked as.
+	// If unset (which is equivalent to setting it to ["<PluginDescript.Name>"]), commands will typically be invocable
```
Comment (marckhouzam) [inline plugin/types.go:109]: `<PluginDescript.Name>` -> `<PluginDescriptor.Name>`

File: plugin/types.go [hunk 3]
```diff
@@ -104,4 +104,19 @@ type PluginDescriptor struct {
...
+	// with the Tanzu CLI using "<PluginDescriptor.Name> <command name> commandargs...."
+	// Can be used to specify additional levels in the command hiearchy via values with space-delimited parts.
```
Comment (marckhouzam) [inline plugin/types.go:111]: `hierarchy`

Comment (marckhouzam) [inline plugin/types.go:111]: Are we really asking for “space-delimited”? I see the field is an array of string

File: plugin/types.go (hunk 1 above)
Comment (marckhouzam) [reply plugin/types.go:112]: This brings up a good point. Do we plan on supporting aliases in the InvokedAs values or must those values be the real name of the command?

Also, what happens if a values in InvokedAs list doesn’t exist?

File: plugin/types.go [hunk 4]
```diff
@@ -104,4 +104,19 @@ type PluginDescriptor struct {
...
+	// e.g. ["ops cluster"] implies plugin's command foo will be invoked by the Tanzu CLI using
+	// 'tanzu ops cluster foo...'
+	// EXPERIMENTAL: subject to change prior to the next official minor release
+	InvokedAs []string `json:"invokedAs,omitempty" yaml:"invokedAs,omitempty"`
+
+	// SupportedContextType specifies one of more ContextType that this plugin will specifically apply to.
```
Comment (marckhouzam) [inline plugin/types.go:117]: “one of more” -> “one or more”

File: plugin/types.go [hunk 5]
```diff
@@ -104,4 +104,19 @@ type PluginDescriptor struct {
...
+	// When no context of matching type is active, the command tree should be omitted by the CLI if possible.
```
Comment (marckhouzam) [inline plugin/types.go:118]: How about “the command tree should be” -> “ the command tree specified by this plugin should be”

File: plugin/types.go [hunk 6]
```diff
@@ -104,4 +104,19 @@ type PluginDescriptor struct {
...
+	// When unset, the plugin does not define any specific opinions on this aspect.
+	// EXPERIMENTAL: subject to change prior to the next official minor release
+	SupportedContextType []string `json:"supportedContextType,omitempty" yaml:"supportedContextType,omitempty"`
```
Comment (marckhouzam) [inline plugin/types.go:121]: Should the type instead be `[]ContextType`?

**Review by marckhouzam — COMMENTED** [review]
A couple of minor clarifications

File: plugin/types.go (hunk 3 above)
Comment (marckhouzam) [reply plugin/types.go:111]: Ok, I think I understand better the intention now.  `InvokedAs` is an array to allow the plugin to create more than one command under the CLI.  But to achieve invocation from deeper in the command-tree, that is when we would have a space-separated value.

Comment (marckhouzam) [reply plugin/types.go:111]: To be clear, are we supporting invoking from deeper in the command-tree? 
E.g., `InvokedAs: []string{"config get details"}` which would create a `details` command under `tanzu config get`?

Comment (vuil) [reply plugin/types.go:111]: yes, but entries in this array is to mapping an entire plugin's command tree to a different location, so your example is unlikely. A more realistic use case could be to use this as an alternate means to associate a plugin's commands under a high level grouping like "operations" (or however we plan to evolve the cmdGroup concept.

File: plugin/types.go (hunk 1 above)
Comment (vuil) [reply plugin/types.go:112]: for simplicity, we only support the real name for now.

> Also, what happens if a values in InvokedAs list doesn’t exist?
This will be explained in the https://github.com/vmware-tanzu/tanzu-cli/pull/687
tldr : that particular remap attempt will fail, but the CLI will likely elect to log and carry on.

**Review by marckhouzam — APPROVED** [review]
LGTM
thanks!
